Improve performance by having better primitives on PSObject

powercode opened the pull request:
By not doing excessive amounts of extra work, formatting can be sped up quite significantly.
The main change comes from adding new, more efficient, primitive to query an object for the existence of an instance member.
The formatting system has been checking for if an object has properties other than some decorated properties added by PS remoting, and it doesn't this by retrieving all properties which results in heavy allocations and wasted cycles.
By adding `GetFirstOrDefault` to `PSObject` and similar primitives to the underlying Adapters, we are able to return early, without having to get all properties back.

Test script:
```powershell
$files = gci -ea:0 c:\windows -Recurse -File
$files | format-table Name, LastAccessTime, Length | out-null
```
version | time | speedup
-- | -- | --
6.1 | 35.5s | 1.0x
PR 8785 | 4.5s | 8x

<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->  

## PR Summary

<!-- Summarize your PR between here and the checklist. -->  

## PR Context  

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.  
- **User-facing changes**
    - [x] Not Applicable
    - **OR**  
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

powercode committed:
[feature] Adding a method to check for an instance member on PSObject.

SteveL-MSFT reviewed (changes requested):
It's going to take a bit of time to review all the changes.  @powercode did you test your changes remotely with WMI and CIM instances?

powercode committed:
Addressing Steve's feedback

powercode commented on src/System.Management.Automation/FormatAndOutput/common/FormatViewManager.cs:
```diff
@@ -464,65 +464,20 @@ internal ViewGenerator ViewGenerator
+            var isRemotingPropertyName =  name.Equals(RemotingConstants.ComputerNameNoteProperty, StringComparison.OrdinalIgnoreCase)
+                   || name.Equals(RemotingConstants.ShowComputerNameNoteProperty, StringComparison.OrdinalIgnoreCase)
+                   || name.Equals(RemotingConstants.RunspaceIdNoteProperty, StringComparison.OrdinalIgnoreCase)
+                   || name.Equals(RemotingConstants.RunspaceIdNoteProperty, StringComparison.OrdinalIgnoreCase)
```
No, just a duplicated line. Removing.

powercode commented on src/System.Management.Automation/FormatAndOutput/common/Utilities/MshObjectUtil.cs:
```diff
@@ -129,7 +54,7 @@ internal static PSPropertyExpression GetDisplayNameExpression(PSObject target, P
             // trying to get something potentially useful
             string[] knownPatterns = new string[] {
-                "name", "id", "key", "*key", "*name", "*id",
+                "name", "id", "key",  "*key", "*name", "*id",
```
fixing

powercode committed:
Fixing code factor issues

powercode committed:
Fixing code factor issues.

powercode committed:
Fixing issue with reversed condition for objects with only remoting properties.

powercode committed:
Ensure we only check properties of the correct type.

powercode commented:
Note to reviewers: @BrucePay @daxian-dbw 

I tried another route in this second PR, by adding another delegate for member lookup. It is less intrusive in the previous functionality, so the risk should be lower.
I wonder if you, when you review this, can try to think about if the same things could be accomplished in a smarter way? I'm adding a way to check for the existence of _any_ property in a reasonably efficient way. But in reality, we only use it to check for very _specific_ properties. (The different stream names, and the remoting properties). In the stream case, it is also important which stream it is, but we don't need a PSNoteProperty for that. For example, the 7 boolean flags in PSObject could become bit fields. Maybe we could encode this information there?

As you can see, I've implemented a special case of a the Linq operator `FirstOrDefault`, but I chose not to make this an iterator as I wanted to keep allocations to a minimum. Not sure it's the right call, just trying to 
explain where I come from :)

iSazonov commented on src/System.Management.Automation/FormatAndOutput/common/FormatViewManager.cs:
```diff
@@ -464,65 +464,19 @@ internal ViewGenerator ViewGenerator
+                   || name.Equals(RemotingConstants.ShowComputerNameNoteProperty, StringComparison.OrdinalIgnoreCase)
+                   || name.Equals(RemotingConstants.RunspaceIdNoteProperty, StringComparison.OrdinalIgnoreCase)
+                   || name.Equals(RemotingConstants.SourceJobInstanceId, StringComparison.OrdinalIgnoreCase)
+                   || name.Equals(RemotingConstants.SourceLength,StringComparison.OrdinalIgnoreCase);
```
There is a chance that `switch` will be faster.

powercode commented:
> It can break remoting/serialization and backward compatibility.
I think that is just an implementation detail. We just need to handle that in the deserializer - but good to keep in mind when doing it.

powercode commented:
I have a sample implementation here - https://github.com/powercode/PowerShell/tree/PSObjectFlags

powercode committed:
PSObjectFlags and WriteStream

Cleaning up PSObject, making the bools into flag fields, leaving room for using the remaing padding for additional flags.

powercode committed:
Fixing code factor issues

powercode commented:
@iSazonov @SteveL-MSFT Any chance we can turn of the hungarian notation warnings from Code Factor? It has nearly 100% false positives.

I don't intend to fix the remaining code factor issues.

SteveL-MSFT commented:
@powercode I believe we're eventually moving off CodeFactor and over to Codacy.  Agree the false positives on hungarian are annoying.

iSazonov commented:
Codacy reports tons of false-positives. We need to merge Dongbo's PR that resolve most CodeFactor issues.

vexx32 commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CustomSerialization.cs:
```diff
@@ -342,7 +342,7 @@ private bool HandlePrimitiveKnownTypePSObject(object source, string property, in
+            if (moSource != null && !moSource.ImmediateBaseObjectIsEmpty)
             {
                 // Check if baseObject is primitive known type
                 object baseObject = moSource.ImmediateBaseObject;
```
Do we need to double check here? Without knowing a great deal about the classes in play here, it looks like you could skip most of the double-checking things in this file and just go right for `object.baseObject = moSource?.ImmediateBaseObject;`

powercode replied:
I agree. Much more elegant. I haven't focused anything on style but tried to keep the changes as small as possible to make it easier to review. Maybe a separate PR?

iSazonov replied:
Fill free to postpone if you have little time now.

vexx32 replied:
@powercode if you'd rather not worry about it, just ping me when this gets merged and I'll do a style PR as a follow up. 🙂

vexx32 commented on src/System.Management.Automation/FormatAndOutput/common/FormatViewManager.cs:
```diff
@@ -550,7 +504,7 @@ internal static bool IsPropertyLessObject(PSObject so)
             {
                 if (DefaultScalarTypes.IsTypeInList(typeNames) ||
-                    IsPropertyLessObject(so))
+                    !HasNonRemotingProperties(so))
```
Common pattern here & elsewhere seems to be to put the boolean operators at the beginning of a continued line.

vexx32 commented on src/System.Management.Automation/FormatAndOutput/common/FormattingObjects.cs:
```diff
@@ -23,6 +23,7 @@
 //
 
 using System.Collections.Generic;
+using System.Management.Automation;
```
Is it necessary to add this?

powercode replied:
Yes, since I moved the WriteStreamType enum

vexx32 commented on src/System.Management.Automation/cimSupport/other/ciminstancetypeadapter.cs:
```diff
@@ -133,6 +133,46 @@ public override PSAdaptedProperty GetProperty(object baseObject, string property
+            {
+                if (cimProperty != null)
+                {
+                    if (predicate(cimProperty.Name))
```
These if statements can be combined.

vexx32 commented on src/System.Management.Automation/engine/COM/ComAdapter.cs:
```diff
@@ -84,6 +84,48 @@ protected override T GetMember<T>(object obj, string memberName)
+                    {
+                        if (lookingForParameterizedProperties)
+                        {
+                            if (predicate.Invoke(prop.Name))
```
These if statements could be combined.

vexx32 commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -3472,6 +3504,43 @@ internal static bool IsTypeParameterizedProperty(Type t)
+                // TODO: check for HiddenAttribute
+                // We can't currently write a parameterized property in a PowerShell class so this isn't too important,
+                // but if someone added the attribute to their C#, it'd be good to set isHidden correctly here.
+                return new PSParameterizedProperty(parameterizedPropertyEntry.propertyName,
```
The common pattern for method declarations spread across multiple lines is to place arguments starting on the next line.
```suggestion
                return new PSParameterizedProperty(
                    parameterizedPropertyEntry.propertyName,
```

vexx32 commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -3504,6 +3573,39 @@ internal static bool IsTypeParameterizedProperty(Type t)
+            }
+
+            MethodBase firstMethod = methodCacheEntry[0].method;
+            var isCtor = firstMethod is ConstructorInfo;
```
Seems like this one could just be evaluated inline in the below expression, I don't see a particular need to have this be a separate variable.

vexx32 commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -4375,6 +4616,8 @@ protected override string ParameterizedPropertyToString(PSParameterizedProperty
 
         #endregion virtual
+
+
```
Extra whitespace here.

vexx32 commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -4687,6 +4972,19 @@ protected override T GetMember<T>(object obj, string memberName)
+        {
+            foreach (var mem in ((PSMemberSet)obj).Members)
+            {
+                if (predicate(mem.Name))
```
```suggestion
                return predicate(mem.Name)) ? mem as T : null;
```
(Remember to replace the body of the if block if you take the suggestion.)

vexx32 commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -4959,6 +5293,18 @@ protected override PSProperty DoGetProperty(object obj, string propertyName)
+        private protected override PSProperty DoGetFirstPropertyOrDefault(object obj, MemberNamePredicate predicate)
+        {
+            XmlNode node = FindFirstNodeOrDefault(obj, predicate);
+            if (node == null)
```
```suggestion
            return (node == null) ? null : new PSProperty(node.LocalName, this, obj, node);
```
Replace the below lines also, if you take the suggestion.

iSazonov replied:
Operator `??`

powercode replied:
Operator ?? is the inversion.

vexx32 commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -111,9 +111,26 @@ internal TypeTable GetTypeTable()
+            {
+                return default;
+            }
+            var member = typeTableToUse.GetFirstOrDefaultMember<T>(msjObj.InternalTypeNames, predicate);
```
We usually have a blank line between a closing brace and subsequent code.

vexx32 commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -194,9 +233,18 @@ internal TypeTable GetTypeTable()
+            // Don't lookup dotnet member if the object doesn't insist.
+            if (msjObj.InternalBaseDotNetAdapter != null)
+            {
+                PSMemberInfo retValue = msjObj.InternalBaseDotNetAdapter.BaseGetMember<PSMemberInfo>(msjObj._immediateBaseObject, predicate);
```
```suggestion
                return msjObj.InternalBaseDotNetAdapter.BaseGetMember<PSMemberInfo>(
                    msjObj._immediateBaseObject, predicate);
```

vexx32 commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -597,56 +638,65 @@ internal static PSObject ConstructPSObjectFromSerializationInfo(SerializationInf
+        private static readonly Collection<CollectionEntry<PSPropertyInfo>> s_propertyCollection = GetPropertyCollection(PSMemberViewTypes.All);
+
+        internal static readonly DotNetAdapter dotNetInstanceAdapter = new DotNetAdapter();
+        private static readonly DotNetAdapter s_baseAdapterForAdaptedObjects = new BaseDotNetAdapterForAdaptedObjects();
```
I'd recommend revising this section so that all `private` and `internal` members are grouped separately, and organised by type and probably alphabetically. There're quite a few members here, so organising this will make it a lot easier to work with in future.

powercode replied:
I think static/instance is more important to group by than accessibility. That is what will show us the object layout. I can make the fields private, and make an internal getter for the fields that need to be accessed internally.

iSazonov replied:
We'll disable CodeFactor rules for such sorting in  #8500 - fill free to feedback.

vexx32 commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -2302,17 +2327,150 @@ DynamicMetaObject IDynamicMetaObjectProvider.GetMetaObject(Expression parameter)
+            {
+                foreach (var instanceMember in _instanceMembers)
+                {
+                    if (instanceMember is PSPropertyInfo propInfo)
```
```suggestion
                    if (instanceMember is PSPropertyInfo propInfo && predicate.Invoke(propInfo.Name))
```

These nested `if`s seem to be mostly unneeded here, I think; collapsing them looks a bit clearer to me.

vexx32 reviewed (commented):
A lot of this is way out of my depth, but I hope some of this is helpful! 😊 

Nothing serious, mostly nitpicks.

iSazonov commented:
@vexx32 I don't look but if you comments is for copy-pasted (not new) code we shouldn't fix it in the PR - it is complicate code review.

powercode committed:
Addressing review comments: Cleanup of PSObject member names and accessibility.

powercode committed:
Addressing Joel's feedback

iSazonov commented on src/System.Management.Automation/engine/COM/ComAdapter.cs:
```diff
@@ -84,6 +84,45 @@ protected override T GetMember<T>(object obj, string memberName)
+                        return new PSParameterizedProperty(prop.Name, this, obj, prop) as T;
+                    }
+
+                    if (lookingForProperties && predicate.Invoke(prop.Name))
```
Seems it's more frequent and could be before previous "if".

powercode replied:
What I've done here is to follow the pattern in the rest of the file. If you look at the context, you will see that there are a lot of the same patterns. That doesn't come across when you only see the diff.

I can change this, but in that case, maybe it should be changed in the whole file?

iSazonov replied:
Oh, not in the PR - it is already huge. (Could we split the PR on some parts? Maybe we could move style/renaming changes in other PR and fast merge before the PR?)

iSazonov commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -597,56 +623,68 @@ internal static PSObject ConstructPSObjectFromSerializationInfo(SerializationInf
+        private PSObjectFlags _flags;
         /// <summary>
-        /// This is the adapter that will depend on the type of baseObject.
+        /// This field contains a stream th
```
Typo?

powercode replied:
yes

iSazonov commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -597,56 +623,68 @@ internal static PSObject ConstructPSObjectFromSerializationInfo(SerializationInf
-                    }
-                }
+        private PSMemberInfoInternalCollection<PSPropertyInfo> _clrMembers;
+        // This is toString value set on deserialization
```
Add new line.

iSazonov commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -597,56 +623,68 @@ internal static PSObject ConstructPSObjectFromSerializationInfo(SerializationInf
+        private PSMemberInfoInternalCollection<PSPropertyInfo> _clrMembers;
+        // This is toString value set on deserialization
+        private string _toStringFromDeserialization;
+        /// <summary>
```
Add new line.

iSazonov commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -672,69 +710,48 @@ internal PSMemberInfoInternalCollection<PSMemberInfo> InstanceMembers
+                if (_adapterSet == null)
                 {
-                    if (psobj._instanceMembers == null)
+                    lock (_lockObject)
```
We could use Lazy

powercode replied:
That would add 24 byte of object header for Lazy. I'm trying to go in the opposite direction: Making PSObject leaner.

iSazonov replied:
Perhaps replacing lock with Interlocked.CompareExchange (Lazy) in PSObject could speed up the operation ~2x. In the file such lock is used in 9 places. For some cmdlets (Select-String, Import-Csv ...) it could be very noticeable.

It would be good to weight performance vs memory. 
Not for the PR.

Another thought. Perhaps there is scenarios in which we could initialize some properties in constructor without lock.

iSazonov commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -2302,17 +2317,173 @@ DynamicMetaObject IDynamicMetaObjectProvider.GetMetaObject(Expression parameter)
+        enum PSObjectFlags : byte
+        {
+            None = 0,
+            /// <summary>
```
Please add new line.
Below too.

powercode replied:
It is a private enum, so there is no "must" here. Not sure what value it would provide.

iSazonov replied:
Please use explicit `private enum PSObjectFlags : byte`.
My previous comment was about bad formatting the code. I expect that there is CodeFactor issues.

iSazonov commented on src/System.Management.Automation/engine/TypeTable.cs:
```diff
@@ -4527,5 +4532,7 @@ private bool ProcessIsBuiltIn(string filePath, ConcurrentBag<string> errors, out
 
         #endregion internal methods
+
+
```
Extra line.

powercode committed:
Fixing typo in comment

powercode committed:
Making the PSObjectFlags enum explicitly private.

powercode committed:
Adding whitespace

daxian-dbw committed:
Remove extra new lines

daxian-dbw reviewed (commented):
I haven't finished even the first pass of the review (just went through `CoreAdapter.cs`) :) 
I submitted a commit to remove some extra new lines I found when quickly browsing the changes. Sent out my partial review comment early.

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -5200,6 +5541,39 @@ private static XmlNode[] FindNodes(object obj, string propertyName, StringCompar
             return retValue.ToArray();
         }
+
+        private static XmlNode FindFirstNodeOrDefault(object obj, MemberNamePredicate predicate)
```
Actually, this might be refactored into a `HasProperty`. The formatting system needs to determine if an object has any property that is not a remoting property. That has previously been done in a very expensive way. When this code was written, I used this method in a place where I needed to find the NoteProperty that determined the output stream, but since that is now refactored, the question we want to have answered is if the psobject has any property that's not one of the magic remoting properties.

powercode replied:
But that is the reason why I want to stop with the first property. I want do to as little work as possible, and stop at the first match. 

But I guess the right thing to do is change it into a boolean test.

powercode replied:
Renaming...

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -1904,6 +1923,19 @@ internal void Add(string name, object member)
+            {
+                if (predicate(entry.Key))
+                {
+                    return entry.Value;
```
Most definitely not :)

powercode replied:
It should be  
```CSharp
return this.memberCollection[entry.Value];                    
```
Love your reviews!

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -3472,6 +3504,43 @@ internal static bool IsTypeParameterizedProperty(Type t)
+                ? GetStaticPropertyReflectionTable((Type)obj)
+                : GetInstancePropertyReflectionTable(obj.GetType());
+
+            object entry = typeTable.GetOrDefault(predicate);
```
Yeah, I got lazy 😔

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -3676,6 +3778,145 @@ protected override T GetMember<T>(object obj, string memberName)
+            return null;
+        }
+
+        private protected T GetFirstPropertyOrDefault<T>(object obj, MemberNamePredicate predicate) where T : PSMemberInfo
```
They do handle the member lookup differently, the latter looping over all members, and the other querying the type table. 
It is beyond me to say if they could be combined.
But the first has
```CSharp
object entry = predicate != null
                               ? typeTable.GetOrDefault(predicate)
                               : typeTable[propertyName];            
```
while the latter has 
```CSharp
    for (int i = 0; i < table.memberCollection.Count; i++)
            {
                var propertyEntry = table.memberCollection[i] as PropertyCacheEntry;
...
```

daxian-dbw replied:
The latter will stop when the entry that satisfy `predicate` is found, and for `typeTable.GetOrDefault(predicate)`, it actually also loop over the `_index` to find the target entry. I think they can be combined.

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -3676,6 +3778,145 @@ protected override T GetMember<T>(object obj, string memberName)
+            {
+                var method = (MethodCacheEntry)table.memberCollection[i];
+                var isCtor = method[0].method is ConstructorInfo;
+                var name = isCtor ? "new" : method[0].method.Name;
```
I don't think you are missing something. I went back and forth when implementing this, and forgot the similarities. Man, I love the way you read code!

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -3676,6 +3778,145 @@ protected override T GetMember<T>(object obj, string memberName)
+        private protected override T GetMember<T>(object obj, MemberNamePredicate predicate)
+        {
+            return GetFirstPropertyOrDefault<T>(obj, predicate) ?? GetFirstMethodOrDefault<T>(obj, predicate) ??
+                   GetFirstEventOrDefault<T>(obj, predicate) ?? GetFirstDynamicMemberOrDefault<T>(obj, predicate);
```
Well, in most cases the predicated version behaves like GetMember**s**, but with a predicate. And the scenario _might_ where it is used, to determine if an object has _any_ property should perhaps find even dynamic properties. 
I have no know use of the events, however.

powercode replied:
fixed

powercode committed:
Addressing Dongbo's feedback.
Changing private protected to protected.
Removing redundancies with shared implementation.
Renaming badly named methods.

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -4756,6 +5054,8 @@ protected override ConsolidatedString GetInternedTypeNameHierarchy(object obj)
         /// <returns>The PSProperty corresponding to propertyName from obj.</returns>
         protected abstract PSProperty DoGetProperty(object obj, string propertyName);
 
+        private protected abstract PSProperty DoGetFirstPropertyOrDefault(object obj, MemberNamePredicate predicate);
```
done

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -4844,6 +5144,40 @@ protected override PSMemberInfoInternalCollection<T> GetMembers<T>(object obj)
+
+            if (typeof(T).IsAssignableFrom(typeof(PSProperty)))
+            {
+                return DoGetFirstPropertyOrDefault(obj, predicate) as T;
```
fixed

powercode commented on src/System.Management.Automation/engine/CoreAdapter.cs:
```diff
@@ -4844,6 +5144,40 @@ protected override PSMemberInfoInternalCollection<T> GetMembers<T>(object obj)
             return returnValue;
         }
+
+        private protected override T GetMember<T>(object obj, MemberNamePredicate predicate)
```
fixed

powercode commented:
@daxian-dbw Please take a moment to see if you can find a better way to accomplish what I try to do. The question we need to answer is if there are any properties that should be rendered. The formatting system excludes the remoting properties by default, so we can think of multiple fast paths. For example, we could store a bool per type (somewhere) caching this result, and only do the heavy lifting when needed. Most CLR types has properties, so we could quickly return "Yes, you have properties to render". If the cache says false, we may still have attached properties, and need to do the more expensive enumeration, but most core scenarios would be improved.

iSazonov commented:
Can our TypeCatalog help for design analyze and the cache?

powercode commented:
@iSazonov Isn't the type catalog just a mapping to the types fullnames? I'm not fully up to speed on this part of the codebase.

iSazonov commented:
@powercode Yes, types to dll-s. My thought was that there is "all" types and we could enhance it with properties to get static cache you need.

daxian-dbw reviewed (commented):
@powercode I finished the first pass of review, and now understand why the overloads with `predicate` are added:
The existing member lookup mechanism allows you to efficiently get back a member if you know the member name, but not when you want a member based on something else. I think this is the right way to go.

Next we need to look into the details. 
I think maybe we should add these two types of overloads to adapters:
```
protected abstract PSMemberInfoInternalCollection<T> GetMembers<T>(object obj, MemberNamePredicate predicate) where T : PSMemberInfo;
protected abstract T GetFirstMemberOrDefault<T>(object obj, MemberNamePredicate predicate) where T : PSMemberInfo;
```
Because later on, it's possible to have a scenario where we need to get all members based on a predicate.

--------------------------------

For the public type `PSPropertyAdapter`, instead of make the new method abstract, let's make it public and virtual:
```
public virtual PSAdaptedProperty GetFirstPropertyOrDefault(object baseObject, MemberNamePredicate predicate)
{
    foreach (var property in GetProperties())
    {
        if (predicate(property.Name))
        ...
    }
}
```
By using the existing `GetProperties()`, we can make a default implementation of `GetFirstPropertyOrDefault`, so it will be backward compatible.

powercode commented on src/System.Management.Automation/engine/ThirdPartyAdapter.cs:
```diff
@@ -323,5 +347,7 @@ public virtual Collection<string> GetTypeNameHierarchy(object baseObject)
         /// </summary>
         public abstract string GetPropertyTypeName(PSAdaptedProperty adaptedProperty);
+
+        internal abstract PSAdaptedProperty GetProperty(object baseObject, MemberNamePredicate propertyName);
```
I think it is just me being conservative with adding a public API - I haven't thought about that scenario. The right thing is probably to make it public.

powercode replied:
fixing

powercode commented on src/System.Management.Automation/FormatAndOutput/common/FormatViewManager.cs:
```diff
@@ -464,65 +464,19 @@ internal ViewGenerator ViewGenerator
 
-            return false;
-        }
+        private static bool HasNonRemotingProperties(PSObject so) => so.GetFirstPropertyOrDefault(NameIsNotRemotingProperty) != null;
```
Yes, that is where most of the speedup comes from. We where getting all properties to see if we had any, and we did so for all objects. Hence my comment above about the possibility to cache this per type.

powercode replied:
That was what started the whole chain of changes. Getting a fast, low-allocation, way of satisfying the demands the formatting system has on the object model. It took up a huge chunk of the total time budget. After these changes, the time is spent doing actual formatting of the objects.

There are probably things to look at there to, but that is for another day :)

powercode commented on src/System.Management.Automation/cimSupport/other/ciminstancetypeadapter.cs:
```diff
@@ -133,6 +133,43 @@ public override PSAdaptedProperty GetProperty(object baseObject, string property
+                    return prop;
+                }
+
+                if (predicate(RemotingConstants.ComputerNameNoteProperty))
```
For sure!

powercode commented on src/System.Management.Automation/engine/ManagementObjectAdapter.cs:
```diff
@@ -198,6 +198,16 @@ protected override T GetMember<T>(object obj, string memberName)
+            // obj should never be null
+            Diagnostics.Assert(obj != null, "Input object is null");
+
+            ManagementBaseObject wmiObject = (ManagementBaseObject)obj;
```
Fixing.

But I wonder if it ever does.

powercode commented on src/System.Management.Automation/engine/ManagementObjectAdapter.cs:
```diff
@@ -872,6 +882,10 @@ internal static string GetMethodDefinition(MethodData mData)
 
+        protected abstract T GetFirstOrDefaultProperty<T>(ManagementBaseObject wmiObject, MemberNamePredicate predicate) where T : PSMemberInfo;
+
+        protected abstract T GetFirstOrDefaultMethod<T>(ManagementBaseObject wmiObject, MemberNamePredicate predicate) where T : PSMemberInfo;
```
I wanted a name that indicates that we don't get all properties that matches the predicate, just the first. Do you think that is expressed in the return type being singular?

daxian-dbw replied:
Agreed. After finishing the whole pass, I realized that. Now I want all the new overloads added are named `GetFirstXXXOrDefault(xxx, MemberNamePredicate predicate)`

powercode commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -4261,6 +4268,19 @@ public override IEnumerator<T> GetEnumerator()
+        {
+            foreach (string name in _members.Keys)
+            {
+                if (predicate.Invoke(name))
```
I have no clue why I used the invoke there. Fixing

powercode commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -4261,6 +4268,19 @@ public override IEnumerator<T> GetEnumerator()
+
+        internal override PSMemberInfo GetFirstOrDefault(MemberNamePredicate predicate)
+        {
+            foreach (string name in _members.Keys)
```
Yes, but it is still the predicated access that is needed. I run the predictate on each name until I find the first match.

I rewrote it with an enumerator to not allocate the Keys collections.

daxian-dbw replied:
Agreed. Now I see why the first-or-default behavior is wanted.

powercode commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -4271,11 +4291,19 @@ internal class CollectionEntry<T> where T : PSMemberInfo
 
-        internal CollectionEntry(GetMembersDelegate getMembers, GetMemberDelegate getMember,
-            bool shouldReplicateWhenReturning, bool shouldCloneWhenReturning, string collectionNameForTracing)
+        internal delegate PSMemberInfo GetFirstOrDefaultDelegate(PSObject obj, MemberNamePredicate predicate);
```
Fixing

powercode commented on src/System.Management.Automation/engine/MshMemberInfo.cs:
```diff
@@ -4902,6 +4932,21 @@ public override IEnumerator<T> GetEnumerator()
 
+        internal override PSMemberInfo GetFirstOrDefault(MemberNamePredicate predicate)
+        {
+            for (int i = 0; i < Collections.Count; i++)
```
The instance members are handled a level above, on PSObject, but I messed the cloning.

Fixing and refactoring.

powercode commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -107,26 +107,59 @@ internal TypeTable GetTypeTable()
+
+        internal static T TypeTableGetFirstOrDefaultMemberDelegate<T>(PSObject msjObj, TypeTable typeTableToUse, MemberNamePredicate predicate) where T : PSMemberInfo
+        {
+            return typeTableToUse?.GetFirstOrDefaultMember<T>(msjObj.InternalTypeNames, predicate);
```
Yes, that is by design. It is unfortunate that the public interface is designed to be so allocation-rich. I try to move in a direction where the allocation of collections are left to the caller as much as possible. The whole point of this exercise is to provide a fast-path that in the best case would not allocate at all - remember that we are only trying to answer the question if a member matching a certain condition exists. No need for any allocations there. 

In general, take an API that is called a 100 000 times for different objects, and each one allocate an empty collection. Compare that to the same API, but this one accepts a collection as a parameter and fills it in with results. The caller of the API knows if it is safe to reuse the collection or not, and in the second case, there is one allocated collection, not 100 000. This is true in many cases. _Leave it to the client_ is a good way to think about it, in my opinion.

The possibility of a `null` return  is also why I have the 'Default' in the name, to indicate that null will be returned if we don't have any matches.

daxian-dbw replied:
That makes perfect sense!

powercode commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -260,6 +299,7 @@ internal static Collection<CollectionEntry<PSMemberInfo>> GetMemberCollection(PS
                 returnValue.Add(new CollectionEntry<PSMemberInfo>(
                     PSObject.DotNetGetMembersDelegate<PSMemberInfo>,
                     PSObject.DotNetGetMemberDelegate<PSMemberInfo>,
+                    PSObject.AdapterGetFirstOrDefaultMemberDelegate<PSMemberInfo>,
```
Yes! I'm so glad you are here! :)

powercode commented on src/System.Management.Automation/engine/MshObject.cs:
```diff
@@ -355,6 +401,7 @@ private static Collection<CollectionEntry<PSMethodInfo>> GetMethodCollection()
                 returnValue.Add(new CollectionEntry<PSPropertyInfo>(
                     PSObject.DotNetGetMembersDelegate<PSPropertyInfo>,
                     PSObject.DotNetGetMemberDelegate<PSPropertyInfo>,
+                    PSObject.AdapterGetFirstOrDefaultMemberDelegate<PSPropertyInfo>,
```
fixing

powercode commented on src/System.Management.Automation/engine/ThirdPartyAdapter.cs:
```diff
@@ -128,6 +128,30 @@ protected override PSProperty DoGetProperty(object obj, string propertyName)
+
+            try
+            {
+                property = _externalAdapter.GetProperty(obj, predicate);
```
Oh, they wont even be able to compile.
That is not ideal!

powercode replied:
fixing

powercode committed:
Addressing Dongbo's comments.

powercode committed:
CodeFactor fixes

daxian-dbw committed:
Rename the new Adapter abstract method to 'GetFirstMemberOrDefault'

powercode commented:
@daxian-dbw I agree completely with your take on `PSPropertyAdapter` and partially with the overload.

If we should add an overload, which I think we should when the need arises, I think it should look like this:
```
protected abstract void GetMembers<T>(
    object obj, 
    MemberNamePredicate predicate, 
    PSMemberInfoInternalCollection<T> resultingMembers)   where T : PSMemberInfo;
```

That way, the collection may be reused by a client with more knowledge about how the it is used. Otherwise we force an allocation of the collection on every call.

You may have the other version too, but implemented as 
```
PSMemberInfoInternalCollection<T> GetMembers<T>(
    object obj, 
    MemberNamePredicate predicate ) where T : PSMemberInfo
{
    var res = new PSMemberInfoInternalCollection<T>();
    GetMembers(obj, predicate, res);
    return res;
}
```
The important thing is that there should always be a variant where the caller can be responsible for the allocations.

daxian-dbw commented:
I completely agree. It's important to make it possible to be allocation free instead of always forcing allocation. We should use the signature you suggested when the need comes, but no need to implement that in this PR.
I actually submitted a PR to your branch https://github.com/powercode/PowerShell/pull/1 to rename the new overloads such as `GetMember`, `DoGetProperty` to `GetFirstXXXOrDefault`.

powercode committed:
Merge pull request #1 from daxian-dbw/myformatting

Rename the new Adapter abstract method to 'GetFirstMemberOrDefault'

powercode commented:
@daxian-dbw  I _really_ like how GitHub enables cooperation on code!

daxian-dbw committed:
Some more fixes and refactor

daxian-dbw commented:
@powercode I submitted another PR to your branch for some fixes and refactoring: https://github.com/powercode/PowerShell/pull/2. Please take a look when you have time.

powercode committed:
Merge pull request #2 from daxian-dbw/myformatting

Some more fixes and refactor

powercode commented:
Looks good!

powercode committed:
Adding back missing namespace

iSazonov commented:
It seems we need rename the PR from
> Faster formatting by having better primitives on PSObject 

to
> Improve performance by having better primitives on PSObject

daxian-dbw reviewed (commented):
LGTM. Now we need to add tests.
We can add xUnit tests to exercise `PSObject.GetFirstPropertyOrDefault`.

Add `[Feature]` tag to the last commit to kick off the feature tests on Linux and macOS.

powercode committed:
[Feature] Adding tests and fixing bux in handling of XmlAttributes

daxian-dbw committed:
Use forward slashes directory separator

daxian-dbw reviewed (approved):
LGTM

iSazonov commented:
Can we move style changes to #9149?

daxian-dbw commented:
@iSazonov I have reviewed all changes in this PR and all look good. Separating the refactoring changes will not only cost extra efforts to resolve conflicts but also require another pass of review to make sure no mistake made during that changes. I don't think it worth the effort.

daxian-dbw commented:
All issues reported by Codacy are false alarms.

iSazonov commented:
>  I don't think it worth the effort.

We agreed to write this in our guide. This allows us to maintain a clear history of functional changes without style ones. Of course it takes effort.
It is not blocking comment.

daxian-dbw commented:
> This allows us to maintain a clear history of functional changes without style ones. Of course it takes effort.

I agree that will get us a clear view of the functional change history. But I think it's more appropriate to bring this up at very early stage of the PR life cycle.
When we reach the end stage of a PR, it's too costly to do this.

iSazonov commented:
> I think it's more appropriate to bring this up at very early stage of the PR life cycle.

At early stage we haven't all style changes. 
From my experience we ready to do this near to finish - and rebase (1) to get new tests, (2) remove unrelated changes.

The main benefit is that the commits are clean.
Otherwise in a few months the details will be forgotten and we will have to figure it out again, not to mention those who will do code analysis in this place for the first time.

> When we reach the end stage of a PR, it's too costly to do this.

In previous powercode's PR you did this :-)

iSazonov commented:
@powercode Thanks for the great improvement!

kborowinski commented:
@powercode This is outstanding work here!
